File: crates/lib/superdev-core/assets/templates/web-react-android-ios-native/libs/native-debug-server-android/src/main/kotlin/com/nativedebugserver/DebugServer.kt
// NativeDebugServer — Lightweight HTTP debug server for Android apps.
// Uses raw ServerSocket for zero-dependency HTTP serving.

package com.nativedebugserver

import android.app.Activity
import android.graphics.Bitmap
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.view.PixelCopy
import java.io.ByteArrayOutputStream
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

/**
 * Lightweight HTTP debug server for exposing app state to AI agents and tools.
 *
 * Usage:
 * ```kotlin
 * DebugServer.start(port = 8080)
 * DebugServer.registerStateProvider("routes") { mapOf("current" to currentRoute) }
 * DebugServer.registerAction("navigate") { params -> navigateTo(params) }
 * DebugServer.log(LogEntry.Level.INFO, "App", "Started")
 * ```
 */
object DebugServer {
    private val logBuffer = LogBuffer()
    private val router = DebugRouter(logBuffer)
    private val server = HttpServer(router)
    private var activity: Activity? = null

    /**
     * Start the debug server. Call from Activity.onCreate() in debug builds.
     * @param port TCP port to listen on (default 8080)
     * @param appId Application identifier (default from BuildConfig)
     * @param activity Activity reference for screenshot capture
     */
    fun start(port: Int = 8080, appId: String = "unknown", activity: Activity? = null) {
        this.activity = activity
        router.configure(appId = appId)

        // Register screenshot provider if activity available
        if (activity != null) {
            router.registerScreenshotProvider { captureScreenshot() }
        }

        server.start(port)
        log(LogEntry.Level.INFO, "DebugServer", "Started on port $port")
    }

    /**
     * Stop the debug server.
     */
    fun stop() {
        server.stop()
        activity = null
        log(LogEntry.Level.INFO, "DebugServer", "Stopped")
    }

    /**
     * Register a state provider called on GET /debug/state/<key>.
     */
    fun registerStateProvider(key: String, provider: () -> Map<String, Any?>) {
        router.registerStateProvider(key, provider)
    }

    /**
     * Register an action handler called on POST /debug/action with matching type.
     */
    fun registerAction(type: String, handler: (Map<String, Any?>) -> Map<String, Any?>) {
        router.registerAction(type, handler)
    }

    /**
     * Register a navigate handler for POST /debug/navigate.
     */
    fun registerNavigateHandler(handler: (String) -> Boolean) {
        router.registerNavigateHandler(handler)
    }

    /**
     * Add a log entry to the buffer.
     */
    fun log(level: LogEntry.Level, tag: String, message: String) {
        logBuffer.append(level, tag, message)
    }

    // MARK: - Screenshot

    @Suppress("DEPRECATION")
    private fun captureScreenshot(): ByteArray? {
        val act = activity ?: return null

        var bitmap: Bitmap? = null
        val latch = CountDownLatch(1)

        Handler(Looper.getMainLooper()).post {
            try {
                val window = act.window
                val view = window.decorView.rootView
                val width = view.width
                val height = view.height

                if (width <= 0 || height <= 0) {
                    latch.countDown()
                    return@post
                }

                val bmp = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    PixelCopy.request(window, bmp, { result ->
                        if (result == PixelCopy.SUCCESS) {
                            bitmap = bmp
                        }
                        latch.countDown()
                    }, Handler(Looper.getMainLooper()))
                } else {
                    view.isDrawingCacheEnabled = true
                    view.buildDrawingCache()
                    val cache = view.drawingCache
                    if (cache != null) {
                        bitmap = cache.copy(Bitmap.Config.ARGB_8888, false)
                    }
                    view.isDrawingCacheEnabled = false
                    latch.countDown()
                }
            } catch (_: Exception) {
                latch.countDown()
            }
        }

        latch.await(3, TimeUnit.SECONDS)

        return bitmap?.let { bmp ->
            val stream = ByteArrayOutputStream()
            bmp.compress(Bitmap.CompressFormat.PNG, 100, stream)
            stream.toByteArray()
        }
    }
}

File: pack/projects/web-react-android-ios-native/libs/native-debug-server-ios/Sources/NativeDebugServer/DebugServer.swift
// NativeDebugServer — Lightweight HTTP debug server for iOS apps.
// Uses Network.framework (NWListener) for zero-dependency HTTP serving.

#if DEBUG
import Foundation
import Network

/// Lightweight HTTP debug server for exposing app state to AI agents and tools.
/// Debug-only — all methods are no-ops in release builds.
///
/// Usage:
/// ```swift
/// DebugServer.shared.start()
/// DebugServer.shared.registerStateProvider("routes") { routeState() }
/// DebugServer.shared.registerAction("navigate") { params in navigate(params) }
/// DebugServer.shared.log(.info, tag: "App", "Started")
/// ```
public final class DebugServer: Sendable {
    public static let shared = DebugServer()

    private let router: DebugRouter
    private let logBuffer: LogBuffer
    private let server: HTTPServer

    private init() {
        self.logBuffer = LogBuffer()
        self.router = DebugRouter(logBuffer: logBuffer)
        self.server = HTTPServer(router: router)
    }

    /// Start the debug server on the given port.
    /// Automatically registers a screenshot provider using ScreenshotCapture.
    public func start(port: UInt16 = 8080, appId: String = Bundle.main.bundleIdentifier ?? "unknown") {
        router.configure(appId: appId)
        registerScreenshotProvider {
            if Thread.isMainThread {
                return MainActor.assumeIsolated { ScreenshotCapture.capturePNG() }
            }
            return DispatchQueue.main.sync {
                MainActor.assumeIsolated { ScreenshotCapture.capturePNG() }
            }
        }
        server.start(port: port)
        log(.info, tag: "DebugServer", "Started on port \(port)")
    }

    /// Stop the debug server.
    public func stop() {
        server.stop()
        log(.info, tag: "DebugServer", "Stopped")
    }

    /// Register a state provider that will be called on `GET /debug/state/<key>`.
    public func registerStateProvider(_ key: String, provider: @escaping @Sendable () -> [String: Any]) {
        router.registerStateProvider(key, provider: provider)
    }

    /// Register an action handler that will be called on `POST /debug/action` with matching type.
    public func registerAction(_ type: String, handler: @escaping @Sendable ([String: Any]) -> [String: Any]) {
        router.registerAction(type, handler: handler)
    }

    /// Register a navigate handler for `POST /debug/navigate`.
    public func registerNavigateHandler(_ handler: @escaping @Sendable (String) -> Bool) {
        router.registerNavigateHandler(handler)
    }

    /// Register a screenshot provider for `GET /debug/screenshot`.
    public func registerScreenshotProvider(_ provider: @escaping @Sendable () -> Data?) {
        router.registerScreenshotProvider(provider)
    }

    /// Add a log entry to the buffer.
    public func log(_ level: LogEntry.Level, tag: String, _ message: String) {
        logBuffer.append(level: level, tag: tag, message: message)
    }
}

#else

// Release stub — all methods are no-ops, zero overhead.
public final class DebugServer: Sendable {
    public static let shared = DebugServer()
    private init() {}

    public func start(port: UInt16 = 8080, appId: String = "") {}
    public func stop() {}
    public func registerStateProvider(_ key: String, provider: @escaping @Sendable () -> [String: Any]) {}
    public func registerAction(_ type: String, handler: @escaping @Sendable ([String: Any]) -> [String: Any]) {}
    public func registerNavigateHandler(_ handler: @escaping @Sendable (String) -> Bool) {}
    public func registerScreenshotProvider(_ provider: @escaping @Sendable () -> Data?) {}
    public func log(_ level: LogEntry.Level, tag: String, _ message: String) {}
}

#endif
